examples/DivisorCurve.jl

```diff
@@ -0,0 +1,327 @@
+end
+
+function AffineCurveDivisor(C::AffinePlaneCurve{S}, P::Point{S}, m::Int=1) where S <: FieldElem
+    iszero(evaluate(C.eq, P.coord)) || error("The point is not on the curve.")
+    return AffineCurveDivisor(C, Dict(P => m))
+end
+
+################################################################################
+
+@doc Markdown.doc"""
+    ProjCurveDivisor(C::ProjPlaneCurve{S}, D::Dict{Oscar.Geometry.ProjSpcElem{S}, Int}) where S <: FieldElem
+
+Given a curve `C` which is assumed to be smooth and irreducible, return the divisor on the curve `C` defined by `D`.
+"""
+mutable struct ProjCurveDivisor{S <: FieldElem} <: CurveDivisor
+    C::ProjPlaneCurve{S}
+    divisor::Dict{Oscar.Geometry.ProjSpcElem{S}, Int}
+    degree::Int
+    function ProjCurveDivisor(C::ProjPlaneCurve{S}, D::Dict{Oscar.Geometry.ProjSpcElem{S}, Int}) where S <: FieldElem
+        for P in keys(D)
+            iszero(evaluate(C.eq, P.v)) || error("The point ", P, "is not on the curve.")
+            if iszero(D[P])
+                delete!(D, P)
+            end
+        end
+        new{S}(C,
+            D,
+            sum(values(D)))
+    end
+end
+#function Base.show(io::IO, D::ProjCurveDivisor)
+#    println(io, "On the curve ", D.C.eq, ", divisor defined by ", D.divisor)
+#end
+
+function ProjCurveDivisor(C::ProjPlaneCurve{S}, P::Oscar.Geometry.ProjSpcElem{S}, m::Int=1) where S <: FieldElem
+    iszero(evaluate(C.eq, P.v)) || error("The point is not on the curve.")
+    return ProjCurveDivisor(C, Dict(P => m))
+end
+
+################################################################################
+# To get a nice output for the divisors.
+
+function AbstractAlgebra.needs_parentheses(A::Array{T, 1}) where T <: RingElem
+    return false
+end
+
+function AbstractAlgebra.expressify(@nospecialize(a::AffineCurveDivisor); context = nothing)
+   prod = Expr(:call, :+)
+   for (p, i) in a.divisor
+      ep = AbstractAlgebra.expressify(p.coord, context = context)
+      push!(prod.args, Expr(:call, :*, i, ep))
+   end
+   return prod
+end
+
+function AbstractAlgebra.needs_parentheses(A::Oscar.Geometry.ProjSpcElem{T}) where T <: RingElem
+    return false
+end
+
+function AbstractAlgebra.expressify(@nospecialize(a::ProjCurveDivisor); context = nothing)
+   prod = Expr(:call, :+)
+   for (p, i) in a.divisor
+      ep = AbstractAlgebra.expressify(p, context = context)
+      push!(prod.args, Expr(:call, :*, i, ep))
+   end
+   return prod
+end
+
+function Base.show(io::IO, ::MIME"text/plain", a::CurveDivisor)
+  print(io, AbstractAlgebra.obj_to_string(a, context = io))
+end
+
+function Base.show(io::IO, a::CurveDivisor)
+  print(io, AbstractAlgebra.obj_to_string(a, context = io))
+end
+
+################################################################################
+################################################################################
+
+function help_add(D::CurveDivisor, E::CurveDivisor)
+    D.C == E.C || error("The divisors are not on the same curve.")
```

Comment:
as an aside: I think there should be a method which returns the curve of a divisor; i.e. instead of writing `D.C` everywhere (thus hardcoding internal details) it would be better to be able to write something like `curve(D)`.

Also, the above check should perhaps be moved into its own helper method, something like `_check_same_curve`; then with the next change I suggest, `help_add` could be replace by two lines (call to `_check_same_curve` and call to `mergewith`)